### PR TITLE
Internal: POC - React history panel [ED-6604]

### PR DESCRIPTION
**Pull request by StyleShit**

### CODE REVIEW

File: packages/editor/examples/src/App.tsx [hunk 1]
```diff
@@ -0,0 +1,73 @@
+import { useState } from 'react';
+import reactLogo from './assets/react.svg';
+import './App.css';
+import './assets/eicons/css/elementor-icons.css';
+import { HistoryPanel, Item } from 'history';
+
```
Comment (matipojo) [inline packages/editor/examples/src/App.tsx:6]: Please try to add `eslint` config to the root of the `editor`

File: packages/editor/examples/src/App.tsx [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+		} );
+	}
+
+	return (
+		<div className="App">
+			<link rel="stylesheet" href="https://connect.elementor.cloud/wp-content/plugins/elementor/assets/css/editor.min.css?ver=3.8.0-cloud1" />
+			{/*<link rel="stylesheet" href="https://connect.elementor.cloud/wp-content/plugins/elementor/assets/css/editor-dark-mode.min.css?ver=3.8.0-cloud1" />*/}
+
+			<h1>With Items:</h1>
+			<HistoryPanel items={ items } currentItem={ currentItem } applyItem={ ( e, args ) => {
```
Comment (matipojo) [inline packages/editor/examples/src/App.tsx:57]:
```suggestion
			<HistoryPanel items={ items } currentItem={ currentItem } onItemClick={ ( e, args ) => {
```

File: packages/editor/examples/src/index.css [hunk 3]
```diff
@@ -0,0 +1,70 @@
+:root {
```
Comment (matipojo) [inline packages/editor/examples/src/index.css:1]: Remove

File: packages/editor/examples/src/vite-env.d.ts [hunk 4]
```diff
@@ -0,0 +1 @@
+/// <reference types="vite/client" />
```
Comment (matipojo) [inline packages/editor/examples/src/vite-env.d.ts:1]: Try to remove

File: packages/editor/package.json [hunk 5]
```diff
@@ -0,0 +1,17 @@
+{
+  "name": "editor",
+  "version": "1.0.0",
+  "description": "",
```
Comment (matipojo) [inline packages/editor/package.json:4]:
```suggestion
```

File: packages/editor/package.json [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+  "main": "index.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "author": "",
+  "license": "ISC",
```
Comment (matipojo) [inline packages/editor/package.json:10]:
```suggestion
  "author": "Elementor Team",
  "license": "GPL3",
```

File: packages/editor/packages/history/.gitignore [hunk 7]
```diff
@@ -0,0 +1,24 @@
+# Logs
```
Comment (matipojo) [inline packages/editor/packages/history/.gitignore:1]: Clean up

File: packages/editor/packages/history/package.json [hunk 8]
```diff
@@ -0,0 +1,35 @@
+{
+  "name": "history",
+  "private": true,
```
Comment (matipojo) [inline packages/editor/packages/history/package.json:3]:
```suggestion
```

File: packages/editor/packages/history/package.json [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    "build": "tsc && vite build",
+    "build:watch": "tsc && vite build --watch",
+    "preview": "vite preview"
+  },
+  "peerDependencies": {
+	"react": "^18.2.0"
+  },
+  "devDependencies": {
+    "@types/react": "^18.0.17",
+    "@types/react-dom": "^18.0.6",
```
Comment (matipojo) [inline packages/editor/packages/history/package.json:17]:
```suggestion
```

File: packages/editor/packages/history/package.json [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+  "version": "0.0.0",
+  "type": "module",
+  "scripts": {
+    "dev": "vite",
+    "build": "tsc && vite build",
+    "build:watch": "tsc && vite build --watch",
+    "preview": "vite preview"
+  },
+  "peerDependencies": {
+	"react": "^18.2.0"
```
Comment (matipojo) [inline packages/editor/packages/history/package.json:13]: eslint

File: packages/editor/packages/history/src/lib/components/HistoryEmpty.tsx [hunk 11]
```diff
@@ -0,0 +1,22 @@
+import useTranslation from '../hooks/useTranslation';
```
Comment (matipojo) [inline packages/editor/packages/history/src/lib/components/HistoryEmpty.tsx:1]: Remove `lib` folder

File: packages/editor/packages/history/src/lib/types/Item.ts [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+
+export type Item = {
+	id: number,
+	status: ItemStatus,
+	title: string,
+	subTitle?: string,
+	action: string,
+}
+
+export type ItemStatus = 'applied' | 'not_applied';
```
Comment (matipojo) [inline packages/editor/packages/history/src/lib/types/Item.ts:11]: Move to top

File: packages/editor/packages/history/src/lib/components/HistoryItem.tsx [hunk 13]
```diff
@@ -0,0 +1,37 @@
+import { ApplyItem, Item } from "../types/Item";
+
+type Props = Item & {
+	onClick: ApplyItem,
+	isCurrent: boolean,
+};
```
Comment (matipojo) [inline packages/editor/packages/history/src/lib/components/HistoryItem.tsx:6]:
```suggestion
type Props = {
	item: Item,
	onClick: ApplyItem,
	isCurrent: boolean,
};
```

File: packages/editor/packages/history/src/lib/components/HistoryItem.tsx [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+		<div
+			className={ `elementor-history-item elementor-history-item-${ props.status } ${ props.isCurrent ? 'elementor-history-item-current' : '' }` }
+			onClick={ ( e : React.MouseEvent<HTMLElement> ) => {
+				props.onClick( e, {
+					id: props.id,
+				} );
+			} }
+		>
+			<div className="elementor-history-item__details">
+				<span className="elementor-history-item__title">{ props.title }</span>{ ' ' }
```
Comment (matipojo) [inline packages/editor/packages/history/src/lib/components/HistoryItem.tsx:19]:
```suggestion
				<span className="elementor-history-item__title">
					{ props.title }
				</span>{ ' ' }
```

File: packages/editor/packages/history/src/lib/components/HistoryItem.tsx [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+				{
+					props.subTitle && (
+						<>
+							<span className="elementor-history-item__subtitle">{ props.subTitle }</span>{ ' '}
```
Comment (matipojo) [inline packages/editor/packages/history/src/lib/components/HistoryItem.tsx:23]:
```suggestion
							<span className="elementor-history-item__subtitle">
								{ props.subTitle }
							</span>{ ' ' }
```

File: packages/editor/packages/history/src/lib/components/HistoryListView.tsx [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+		<>
+			{
+				props.items.map( ( item, i ) =>
+					<HistoryItem
+						key={ item.id }
+						id={ item.id }
+						title={ item.title }
+						subTitle={ item.subTitle }
+						action={ item.action }
+						status={ item.status }
```
Comment (matipojo) [inline packages/editor/packages/history/src/lib/components/HistoryListView.tsx:22]: Item

File: packages/editor/packages/history/src/lib/components/HistoryPanel.tsx [hunk 17]
```diff
@@ -0,0 +1,37 @@
+import HistoryEmpty from './HistoryEmpty';
+import HistoryListView from './HistoryListView';
+import React from "react";
+import { ApplyItem, Item } from "../types/Item";
+import useTranslation from "../hooks/useTranslation";
+
+type Props = {
+	items: Item[],
+	currentItem: number,
+	applyItem: ApplyItem,
```
Comment (matipojo) [inline packages/editor/packages/history/src/lib/components/HistoryPanel.tsx:10]:
```suggestion
	onItemClick: ApplyItem,
```

File: packages/editor/packages/history/src/lib/vite-env.d.ts [hunk 18]
```diff
@@ -0,0 +1 @@
+/// <reference types="vite/client" />
```
Comment (matipojo) [inline packages/editor/packages/history/src/lib/vite-env.d.ts:1]: Remove

**Review by matipojo — CHANGES_REQUESTED** [review]

File: packages/editor/examples/tsconfig.json [hunk 19]
```diff
@@ -0,0 +1,21 @@
+{
```
Comment (matipojo) [inline packages/editor/examples/tsconfig.json:1]: Try to import from root

**Review by matipojo — CHANGES_REQUESTED** [review]

File: packages/editor/examples/.gitignore [hunk 20]
```diff
@@ -0,0 +1,24 @@
+# Logs
```
[the remark this answers is not in the document]
Comment (StyleShit) [reply packages/editor/examples/.gitignore:1]: Done

File: packages/editor/examples/package-lock.json [hunk 21]
```diff
@@ -0,0 +1,2358 @@
+{
```
[the remark this answers is not in the document]
Comment (StyleShit) [reply packages/editor/examples/package-lock.json:1]: Done

File: packages/editor/examples/index.html [hunk 22]
```diff
@@ -0,0 +1,12 @@
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <meta charset="UTF-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>History Panel</title>
```
[the remark this answers is not in the document]
Comment (StyleShit) [reply packages/editor/examples/index.html:6]: Done

File: packages/editor/examples/package.json [hunk 23]
```diff
@@ -0,0 +1,23 @@
...
+  "type": "module",
+  "scripts": {
+    "dev": "vite",
+    "build": "tsc && vite build",
+    "preview": "vite preview"
+  },
+  "dependencies": {
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0",
+	"history": "*"
```
[the remark this answers is not in the document]
Comment (StyleShit) [reply packages/editor/examples/package.json:14]: Done

File: packages/editor/examples/src/App.tsx [hunk 24]
```diff
@@ -0,0 +1,73 @@
+import { useState } from 'react';
+import reactLogo from './assets/react.svg';
```
[the remark this answers is not in the document]
Comment (StyleShit) [reply packages/editor/examples/src/App.tsx:2]: Done

File: packages/editor/examples/src/App.tsx (hunk 2 above)
Comment (StyleShit) [reply packages/editor/examples/src/App.tsx:57]: Done

File: packages/editor/examples/src/index.css (hunk 3 above)
Comment (StyleShit) [reply packages/editor/examples/src/index.css:1]: Done

File: packages/editor/examples/src/vite-env.d.ts (hunk 4 above)
Comment (StyleShit) [reply packages/editor/examples/src/vite-env.d.ts:1]: Done

File: packages/editor/package.json (hunk 5 above)
Comment (StyleShit) [reply packages/editor/package.json:4]: Done

File: packages/editor/package.json (hunk 6 above)
Comment (StyleShit) [reply packages/editor/package.json:10]: Done

File: packages/editor/packages/history/.gitignore (hunk 7 above)
Comment (StyleShit) [reply packages/editor/packages/history/.gitignore:1]: Done

File: packages/editor/packages/history/package.json (hunk 8 above)
Comment (StyleShit) [reply packages/editor/packages/history/package.json:3]: https://docs.npmjs.com/cli/v8/configuring-npm/package-json#private

Comment (StyleShit) [reply packages/editor/packages/history/package.json:3]: It depends on how we're gonna publish it (as multiple packages or as a whole "Editor" pacakge):

https://docs.npmjs.com/cli/v8/configuring-npm/package-json#private

File: packages/editor/packages/history/package.json (hunk 9 above)
Comment (StyleShit) [reply packages/editor/packages/history/package.json:17]: Done

File: packages/editor/packages/history/package.json (hunk 10 above)
Comment (StyleShit) [reply packages/editor/packages/history/package.json:13]: Done

File: packages/editor/packages/history/src/lib/components/HistoryEmpty.tsx (hunk 11 above)
Comment (StyleShit) [reply packages/editor/packages/history/src/lib/components/HistoryEmpty.tsx:1]: Done

File: packages/editor/packages/history/src/lib/types/Item.ts (hunk 12 above)
Comment (StyleShit) [reply packages/editor/packages/history/src/lib/types/Item.ts:11]: Done

File: packages/editor/packages/history/src/lib/components/HistoryItem.tsx (hunk 13 above)
Comment (StyleShit) [reply packages/editor/packages/history/src/lib/components/HistoryItem.tsx:6]: Done

File: packages/editor/packages/history/src/lib/components/HistoryItem.tsx (hunk 14 above)
Comment (StyleShit) [reply packages/editor/packages/history/src/lib/components/HistoryItem.tsx:19]: Done

File: packages/editor/packages/history/src/lib/components/HistoryItem.tsx (hunk 15 above)
Comment (StyleShit) [reply packages/editor/packages/history/src/lib/components/HistoryItem.tsx:23]: Done

File: packages/editor/packages/history/src/lib/components/HistoryListView.tsx (hunk 16 above)
Comment (StyleShit) [reply packages/editor/packages/history/src/lib/components/HistoryListView.tsx:22]: Done

File: packages/editor/packages/history/src/lib/components/HistoryPanel.tsx (hunk 17 above)
Comment (StyleShit) [reply packages/editor/packages/history/src/lib/components/HistoryPanel.tsx:10]: Done

File: packages/editor/packages/history/src/lib/vite-env.d.ts (hunk 18 above)
Comment (StyleShit) [reply packages/editor/packages/history/src/lib/vite-env.d.ts:1]: Done

File: packages/editor/.eslintrc.js [hunk 25]
```diff
@@ -0,0 +1,47 @@
+module.exports = {
+	// Extends: [
+	// 	'plugin:@typescript-eslint/recommended',
+	// 	'plugin:react/recommended',
+	// 	'plugin:@wordpress/eslint-plugin/recommended-with-formatting',
+	// ],
+	// plugins: [
+	// 	'react',
+	// 	'@typescript-eslint',
+	// ],
```
Comment (matipojo) [inline packages/editor/.eslintrc.js:10]: ?

**Review by matipojo — CHANGES_REQUESTED** [review]

File: packages/editor/.eslintrc.js (hunk 25 above)
Comment (StyleShit) [reply packages/editor/.eslintrc.js:10]: WIP

File: packages/editor/examples/src/assets/eicons/css/elementor-icons.css [hunk 26]
```diff
@@ -0,0 +1,1363 @@
+@font-face {
+  font-family: 'eicons';
+  src: url("../fonts/eicons.eot?5.16.0");
+  src: url("../fonts/eicons.eot?5.16.0#iefix") format("embedded-opentype"), url("../fonts/eicons.woff2?5.16.0") format("woff2"), url("../fonts/eicons.woff?5.16.0") format("woff"), url("../fonts/eicons.ttf?5.16.0") format("truetype"), url("../fonts/eicons.svg?5.16.0#eicon") format("svg");
```
Comment (matipojo) [inline packages/editor/examples/src/assets/eicons/css/elementor-icons.css:4]: Remove unnecessary formats

File: packages/editor/examples/src/App.tsx [hunk 27]
```diff
@@ -0,0 +1,73 @@
+import { useState } from 'react';
+import './App.css';
+import './assets/eicons/css/elementor-icons.css';
+import { HistoryPanel, Item } from 'history';
```
Comment (matipojo) [inline packages/editor/examples/src/App.tsx:4]: Disable Rule

File: packages/editor/.eslintrc.js [hunk 28]
```diff
@@ -0,0 +1,48 @@
...
+		'comma-dangle': [ 'error', 'always-multiline' ],
+		'no-undef': 'off',
+		'dot-notation': 'error',
+		'no-shadow': 'error',
+		'no-lonely-if': 'error',
+		'no-mixed-operators': 'error',
+		'no-nested-ternary': 'error',
+		'no-cond-assign': 'error',
+		'no-unused-vars': 'off', // Using the @typescript-eslint/no-unused-vars rule instead.
+		indent: [ 'off', 'tab', { SwitchCase: 1 } ],
```
Comment (matipojo) [inline packages/editor/.eslintrc.js:25]:
```suggestion
```

File: packages/editor/.eslintrc.js [hunk 29]
```diff
@@ -0,0 +1,48 @@
...
+	],
+	plugins: [
+		'react',
+		'@typescript-eslint',
+	],
+	parser: '@typescript-eslint/parser',
+	rules: {
+		'computed-property-spacing': [ 'error', 'always' ],
+		'comma-dangle': [ 'error', 'always-multiline' ],
+		'no-undef': 'off',
```
Comment (matipojo) [inline packages/editor/.eslintrc.js:17]:
```suggestion
```

File: packages/editor/examples/src/vite-env.d.ts (hunk 4 above)
Comment (matipojo) [reply packages/editor/examples/src/vite-env.d.ts:1]: Add a comment about keeping the file

File: packages/editor/packages/history/src/components/HistoryItem.tsx [hunk 30]
```diff
@@ -0,0 +1,47 @@
...
+type Props = {
+	item: Item,
+	onClick: OnItemClick,
+	isCurrent: boolean,
+};
+
+const HistoryItem : React.VFC<Props> = ( props ) => {
+	return (
+		// eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions
+		<div
```
Comment (matipojo) [inline packages/editor/packages/history/src/components/HistoryItem.tsx:12]: Try button

**Review by matipojo — CHANGES_REQUESTED** [review]

File: packages/editor/examples/src/vite-env.d.ts (hunk 4 above)
Comment (StyleShit) [reply packages/editor/examples/src/vite-env.d.ts:1]: Done

Comment (StyleShit) [reply packages/editor/examples/src/vite-env.d.ts:1]: Done

File: packages/editor/examples/tsconfig.json (hunk 19 above)
Comment (StyleShit) [reply packages/editor/examples/tsconfig.json:1]: Done

File: packages/editor/examples/src/assets/eicons/css/elementor-icons.css (hunk 26 above)
Comment (StyleShit) [reply packages/editor/examples/src/assets/eicons/css/elementor-icons.css:4]: Done

File: packages/editor/examples/src/App.tsx (hunk 27 above)
Comment (StyleShit) [reply packages/editor/examples/src/App.tsx:4]: Done

File: packages/editor/.eslintrc.js (hunk 28 above)
Comment (StyleShit) [reply packages/editor/.eslintrc.js:25]: Done

File: packages/editor/.eslintrc.js (hunk 29 above)
Comment (StyleShit) [reply packages/editor/.eslintrc.js:17]: Done

Comment (StyleShit) [reply packages/editor/.eslintrc.js:17]: Done

File: packages/editor/packages/history/src/components/HistoryItem.tsx (hunk 30 above)
Comment (StyleShit) [reply packages/editor/packages/history/src/components/HistoryItem.tsx:12]: Breaks the style:


image

File: packages/editor/packages/history/package.json [hunk 31]
```diff
@@ -0,0 +1,33 @@
+{
```
Comment (matipojo) [inline packages/editor/packages/history/package.json:1]: Make it a template `editor/packages/create-template/template`

File: packages/editor/examples/src/App.tsx [hunk 32]
```diff
@@ -0,0 +1,74 @@
+import { useState } from 'react';
+import './App.css';
+import './assets/eicons/css/elementor-icons.css';
+// eslint-disable-next-line import/no-unresolved
+import { HistoryPanel, Item } from 'history';
```
Comment (matipojo) [inline packages/editor/examples/src/App.tsx:5]: import { HistoryActionsPanel, HistoryActionsItem } from 'history-actions';

**Pull request merged**